compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,645 @@
+            context, std::array<int64_t, 3>{0, 2, 1});
+    sinkingPatterns
+        .insert<NamedOpConversion</*OpType=*/linalg::BatchMatmulTransposeAOp,
+                                  /*ReplacementType=*/linalg::BatchMatmulOp,
+                                  /*inputIdx=*/0,
+                                  /*inputRank=*/3>>(
+            context, std::array<int64_t, 3>{0, 2, 1});
+    sinkingPatterns.insert<SinkTransposeThroughExtractSlice>(context);
+    sinkingPatterns.insert<SinkTransposeThroughExpandShape>(context);
+    sinkingPatterns.insert<FuseTransposeWithGenericConsumer>(context);
+    if (failed(
+            applyPatternsAndFoldGreedily(funcOp, std::move(sinkingPatterns)))) {
+      return signalPassFailure();
+    }
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "\n--- After sinking transpose ops down ---\n";
+    funcOp->print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+    llvm::dbgs() << "\n\n";
+  });
+
+  // Pre-empting additional patterns for bubbling transposes.
+  // if (!testSinkingOnly) {
+  //   ...
+  // }
+
+  // Currently this only runs after all propagation has finished. There are
+  // cases where combining transposes can allow further propagation, but
+  // similarly there are cases where combining adjacent transposes limits later
+  // propagation patterns. For now this keeps it simple as once propagation has
+  // finished, it should in all cases be better to fuse.
+  // TODO: Run this to some kind of fixed point with propagation. This is tricky
+  // because propagation can make trivial modifications to the IR (e.g. through
+  // reshapes).
+  {
+    RewritePatternSet patterns(context);
+    patterns.insert<ComposeTransposes>(context);
+    if (enableAggressivePropagation) {
+      patterns.insert<GeneralizeInputTransposedNamedOp>(context);
+    }
+    if (failed(applyPatternsAndFoldGreedily(funcOp, std::move(patterns)))) {
+      return signalPassFailure();
+    }
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "\n--- After combining transpose ops ---\n";
+    funcOp->print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+    llvm::dbgs() << "\n\n";
+  });
+
+  // Re-generalize any remaining transposes. Later pipelines expect it.
+  if (!keepTransposes) {
+    SmallVector<linalg::LinalgOp> transposeCandidates;
+    funcOp.walk([&](linalg::LinalgOp linalgOp) {
+      if (!IREE::Flow::isNonNullAndOutsideDispatch(linalgOp)) {
+        return;
+      }
+      if (isa_and_nonnull<linalg::TransposeOp>(linalgOp.getOperation())) {
+        transposeCandidates.push_back(linalgOp);
+      }
+    });
+    IRRewriter rewriter(&getContext());
+    for (auto linalgOp : transposeCandidates) {
+      rewriter.setInsertionPoint(linalgOp);
+      FailureOr<linalg::GenericOp> generalizedOp =
+          linalg::generalizeNamedOp(rewriter, linalgOp);
+      if (failed(generalizedOp)) {
+        linalgOp->emitOpError("failed to generalize operation");
+        return signalPassFailure();
+      }
+    }
+  }
```

Comment:
I thought it is done in the later passes. We can just move it out of `if (transformOptions.options.dataTiling)` check, then we don't need these.

https://github.com/openxla/iree/blob/6c1b1ab6ca2d628570d779a775a82ff4662a6a3f/compiler/src/iree/compiler/GlobalOptimization/Passes.cpp#L110-L119